### PR TITLE
Add Plots recipe

### PR DESCRIPTION
**Pull request by mkborregaard**
WIP - work in progress

Adds plotting functionality to shapefiles by a recipe for the Plots.jl package. A lot of the code in the PR is written by Tom Breloff (the owner of Plots.jl).
Right now only works for Polygon shapefiles (type 5) but I will expand this if you are interested.

An example that works:
```
# load the packages
using Shapefile, Plots
pyplot()

# get an example shapefile
using Requests #Just to get the example shapefile
tmp = get("https://github.com/nvkelso/natural-earth-vector/raw/master/110m_physical/ne_110m_land.shp")
save(tmp, "tmp.shp")
shp = open("tmp.shp") do fd
    read(fd, Shapefile.Handle)
end;

#plot it 
plot(shp, color = :green, aspect_ratio = 1, legend = false)

### CODE REVIEW

File: src/Shapefile.jl [hunk 1]
```diff
@@ -1,30 +1,30 @@
 module Shapefile
     import Base: read, show, +, /, ./
 
-    export Handle, RGBGradient, LogLinearRGBGradient, LinearRGBGradient
- 
+    export Handle, RGBGradient, LogLinearRGBGradient, LinearRGBGradient, apply_recipe
```
Comment (tbreloff) [inline src/Shapefile.jl]: This is unnecessary

File: src/Plotsrecipe.jl [hunk 2]
```diff
@@ -0,0 +1,40 @@
+using RecipesBase
```
Comment (tbreloff) [inline src/Plotsrecipe.jl]: You either have to make this conditional like the compose check, or add RecipesBase to the REQUIRE file

**Comment by yeesian** [comment]
@penntaylor might this be of interest to you w.r.t. your work on https://github.com/penntaylor/ChoroplethMaps.jl?

**Comment by mkborregaard** [comment]
Huh. Just worked, but now it appears broken - just as I came home to Friday evening with the kids. Will fix over the weekend.

**Comment by tbreloff** [comment]
@mkborregaard I just had the thought that, if you were happy to treat every `ESRIShape` as a separate series, then you could remove the vector and matrix versions and replace with an array version, and the issue of "keeping polygon parts together" would be magically solved.  Here's the new recipes:

```julia
function shapefile_coords(poly::Shapefile.ESRIShape)
    start_indices = poly.parts+1
    end_indices = vcat(poly.parts[2:end], length(poly.points))
    x, y = zeros(0), zeros(0)
    for (si,ei) in zip(start_indices, end_indices)
        push!(x, NaN)
        push!(y, NaN)
        for pt in poly.points[si:ei]
            push!(x, pt.x)
            push!(y, pt.y)
        end
    end
    x, y
end

function shapefile_coords{T<:Shapefile.ESRIShape}(polys::AbstractArray{T})
    x, y = [], []
    for poly in polys
        xpart, ypart = shapefile_coords(poly)
        push!(x, xpart)
        push!(y, ypart)
    end
    x, y
end

@recipe f(poly::Shapefile.ESRIShape) = (seriestype --> :shape; shapefile_coords(poly))
@recipe f{T<:Shapefile.ESRIShape}(polys::AbstractArray{T}) = (seriestype --> :shape; shapefile_coords(polys))
```

tmp

**Comment by yeesian** [comment]
I suggest we identify (i) what elements of this PR should go into `https://github.com/JuliaPlots/*`, and (ii) what elements of it is more generically tied to extracting coordinates from ESRIShapes. If that requires too much work upfront, it might be better to hash it out in a package like https://github.com/penntaylor/ChoroplethMaps.jl first.

Having [gone down this path with GeoJSON](https://github.com/JuliaGeo/GeoJSON.jl/tree/geo-interface) before, I personally think the task of extracting coordinates/etc was meant to be addressed by a more general/"base" package like https://github.com/JuliaGeo/GeoInterface.jl

**Comment by tbreloff** [comment]
@yeesian I personally feel like these belong in Shapefile.jl, because they are simple conversions of a Shapefile type to the underlying coordinates.  There's no graphics/plotting, and there's nothing generic about it.  The idea behind RecipesBase is that it's a super lightweight base class that lets you "plug in" to the complex Plots framework without actually adding the dependency.  All you do is define a conversion to a lower-level type... in this case we convert `Shapefile.ESRIShape` to `Vector{Float64}`s and we're done.  Colors, strokes, overlayed annotation, and whatever else you can think of are instantly available.

**Comment by yeesian** [comment]
> in this case we convert `Shapefile.ESRIShape` to `Vector{Float64}`

I very much want to agree, but you could also convert it to `Matrix{Float64}` or `Vector{Tuple{Float64,Float64}}` instead of `Tuple{Vector{Float6},Vector{Float6}}` too. I don't want my views to be a blocker to otherwise good progress though; so I'll be okay if any of the other JuliaGeo members would like to merge it.

> The idea behind RecipesBase is that it's a super lightweight base class that lets you "plug in" to the complex Plots framework without actually adding the dependency.

I understand the intent, and we can do it as an experiment, but I'm still unconvinced about the dependency on RecipesBase. I'll personally prefer to see it work out into a mature solution for a proxy package in the meantime before merging it into this package though.

**Comment by mkborregaard** [comment]
@tbreloff That is fantastic, I was musing over the same lines, but as usual you are way ahead.

**Comment by mkborregaard** [comment]
> I suggest we identify (i) what elements of this PR should go into https://github.com/JuliaPlots/*

So we actually did this - the code I and (mainly) Tom wrote has been divided in two, this PR consists of stuff that appear most relevant to put into JuliaGeo, and the other part will be a PR to https://github.com/JuliaPlots/PlotRecipes.jl , allowing for more tailored shapefile plots.
Tom's idea, the way I understand it, is to allow plotting support for packages in the most lightweight way possible, given julia's distributed plotting system – just the very light dependency RecipesBase and about a page or less of code.

@yeesian Do you want to make this a PR to GeoInterface.jl instead, using the interface within that package? I can see how that would make a lot of sense and avoid duplicating functionality. I was not aware of the intended package structure for JuliaGeo use, so I have just used Shapefiles.jl. Like most users, I guess, I do not have GeoInterface installed. Are you intending to put all this in a metapackage like https://github.com/BioJulia/Bio.jl?

I would tend to think that users would expect shapefiles to plot out of the box. This is my reason for the PR – I am not affiliated with Plots in any way, I am just trying to support a julia package ecosystem I can use in my own research and recommend to students. I can think of very few uses of shapefiles that do not involve plotting - if for nothing else then simply to check that operations have had the intended results. The lack of (or poor) plotting functionality for shapefiles has kept me in R for all my shapefile work, while I have transitioned to julia for many other tasks. This was my reason to comment on an issue at Plots, asking for @tbreloff 's help in creating this functionality for shapefiles, resulting in this PR.

**Comment by mkborregaard** [comment]
I should add that I see two advantages to having this under JuliaGeo rather than under JuliaPlots:

1. If the shapefile format gets updated (it is currently quite rudimentary, right?) it will be easier to update this functionality along with it.
2. If another competing type of shapefile appear, it will be easy for Plots to support this as well. And at the same type, the code in this PR will make it easy for JuliaGeo to support other plotting packages. Useful in a package system developing so quickly as Julia's.

**Comment by yeesian** [comment]
> Do you want to make this a PR to GeoInterface.jl instead, using the interface within that package?

That was my original intention, but I haven't pushed on it, since I don't have a good sense of the range of basic methods such an interface should support. I think a richer vocabulary in describing the "grain" of different geometries* (similar to the distinction between row vs column-based tables) will be needed first. The [effort in wrapping GDAL](https://github.com/visr/GDAL.jl) has been helpful in eliciting what they might be. Until then, I don't think it's a good idea for any package to depend on GeoInterface, which is why it has not showed up anywhere.

*across GeoJSON, Shapefile, GDAL, LibGEOS, Geodesy, Proj4, etc

> I would tend to think that users would expect shapefiles to plot out of the box.

Such users might not have to use this package though. And I'm still not sure an accumulation of shims will get us there.

> I am just trying to support a julia package ecosystem I can use in my own research and recommend to students.

I recommend starting a package (requiring Shapefile, RecipeBase, etc) that does what you want in this PR. It will be a lot easier for you to push on your ideas, without having to "ask for permission" through multiple rounds of reviews. You don't have to punt on a more general solution, if you have specific use cases in mind, and can already see a specific path ahead for achieving them.

**Comment by tbreloff** [comment]
> I recommend starting a package (similar to ChloroplethMaps; but requiring Shapefile, RecipeBase, etc) that does what you want in this PR.

@yeesian Thanks for your thoughts.  I do understand your hesitation, as I would have the same hesitation if I wasn't so actively involved with Plots.  I think you'll come around eventually, but in the meantime, I can add these recipes to PlotRecipes with a conditional inclusion when Shapefile.jl is installed.  I don't like all these conditional inclusions, so I hope eventually we can get these recipes as close to the source as possible.

As you suggested, it might make sense to include related functionality from similar packages in one place, wrapping downloading the shape data, recipes, etc.  For now I'm trying to keep package explosion to a minimum, while still modularizing in a nice way.

**Comment by yeesian** [comment]
>  I think you'll come around eventually, but in the meantime, I can add these recipes to PlotRecipes with a conditional inclusion when Shapefile.jl is installed.

Thanks! I'm impressed by the pace of development in Plots and its recipes, and am hopeful I will too. I don't like conditional dependencies, and am sorry we don't yet have the kind of solution that GeoInterface aspires to.

> For now I'm trying to keep package explosion to a minimum, while still modularizing in a nice way.

It will be easier to identify mature elements from other packages worth emulating, than to get the desired degree of modularization right the first time. I'm less worried about package explosion outside of this organization, and am more concerned that we lack the maintainers for the features we want to support in this organization.

**Comment by mkborregaard** [comment]
> I can add these recipes to PlotRecipes

Thanks a lot @tbreloff . This is going to make life so much easier for me and anyone else who wants to use shapefiles to create maps in Julia. Do you want me to fork and write the code for the other shapefile types? Polygons with holes is an unsolved issue.

> I recommend starting a package (similar to ChloroplethMaps; but requiring Shapefile, RecipeBase, etc) that does what you want in this PR

This was exactly what I wanted - a plotting function for shapefiles. I will follow the further progress of JuliaGeo with great interest.

Thanks to both for your hard work making the tools for tomorrow.

**Pull request closed without merging**